### PR TITLE
Add a rate limit for password resets

### LINKED ISSUE #7502: Rate Limit absent on forgot password page
### Affected version(s)

all

### Description
This Vulnerability can be used as an email bomber for bombing some user accounts with forgotten password emails. Also, the Impact is high on mail servers. If some attacker sends emails using the same loop, It will keep the mail servers busy and can make them unavailable to other users.

STEPS TO REPRODUCE:
1) Go to password reset password enter email and intercept the request in burp.
2) Now send that request to Intruder and select payloads for user-agent.
3) Start the attack
4) Now check the email address is been flooded with mail

### PR DESCRIPTION
**Pull request by bytehead**
Fixes #7502

### CODE REVIEW

**Review by aschempp — COMMENTED** [review]
this would only limit the number of attempts of **one** user, right? But I could still randomly "hack" and submit 1000 different user names to spam them with reset links?

**Comment by leofeyer** [comment]
You can now send a maximum of three mails per user. Previously there was no limit. This is the problem that this PR fixes.

**Comment by bytehead** [comment]
Yes. You could, but I think you would have to guess existing user and/or email combinations otherwise no email is sent.

**Comment by aschempp** [comment]
> You can now send a maximum of three mails per user. Previously there was no limit. This is the problem that this PR fixes.

within 24 hours, right? because we obviously can't limit that "forever" 😅

File: core-bundle/src/Resources/contao/models/OptInModel.php [hunk 1]
```diff
@@ -166,6 +166,28 @@ public static function findByRelatedTableAndIds($strTable, array $arrIds, array
...
+	/**
+	 * Count unconfirmed password reset opt-in tokens by their ID
+	 *
+	 * @param array $arrIds
+	 *
+	 * @return int
+	 *
+	 * @throws \Exception
+	 */
+	public static function countUnconfirmedPasswordResetTokensByIds(array $arrIds)
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/models/OptInModel.php]:
```suggestion
	public static function countUnconfirmedPasswordResetTokensById($intId)
```

Why use an array if we ever only provide one ID?

Comment (bytehead) [reply core-bundle/src/Resources/contao/models/OptInModel.php]: Adjusted in 6f61ac89cd.

**Review by leofeyer — DISMISSED** [review]
The code works fine, but I was just wondering if 24 hours is too long a period? If you as a user find yourself in this situation, there is nothing you can do but wait. Maybe one hour would be enough already?

File: core-bundle/src/Resources/contao/languages/en/default.xlf [hunk 2]
```diff
@@ -1910,6 +1910,9 @@
       <trans-unit id="MSC.invalidTokenUrl">
         <source>The link you were trying to open could not be verified. If you have clicked the link yourself or have received it by a trustworthy person, you can confirm the process below.</source>
       </trans-unit>
+      <trans-unit id="MSC.tooManyPasswordResetAttempts">
+        <source>You have tried to reset your password too many times.</source>
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/languages/en/default.xlf]:
```suggestion
        <source>You have tried to reset your password too many times. Please try again in 24 hours.</source>
```

So the user knows what happens next?

**Comment by bytehead** [comment]
> Maybe one hour would be enough already?

@ausi what's your opinion on this?

**Comment by ausi** [comment]
> > Maybe one hour would be enough already?
> 
> @ausi what's your opinion on this?

As we only allow three attempts, I think 15 minutes would be enough.

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bytehead.

**Pull request merged**